cpp/src/dataframe/device_table.cuh

```diff
@@ -0,0 +1,543 @@
+    row_valid_mask = ~(row_valid_mask);
+
+    for (gdf_size_type i = 0; i < _num_columns; ++i) {
+      const gdf_valid_type* current_column_mask = column_valid_masks[i];
+
+      // The column validity mask is optional and can be nullptr
+      if (nullptr != current_column_mask) {
+        row_valid_mask &= current_column_mask[mask_number];
+      }
+    }
+    return row_valid_mask;
+  }
+
+  const gdf_size_type _num_columns;
+  gdf_valid_type** column_valid_masks;
+};
+
+
+}  // namespace
+
+/** 
+ * @brief Provides row-level device functions for operating on a set of columns.
+ */
+class device_table : public managed
+{
+public:
+
+  using size_type = int64_t;
+
+  /**---------------------------------------------------------------------------*
+   * @brief Factory function to construct a device_table wrapped in a
+   * unique_ptr.
+   *
+   * Constructing a `device_table` via a factory function is required to ensure
+   * that it is constructed via the `new` operator that allocates the class with
+   * managed memory such that it can be accessed via pointer or reference in
+   * device code. A `unique_ptr` is used to ensure the object is cleaned-up correctly.
+   * 
+   * Usage:
+   * ```
+   * gdf_column * col;
+   * auto device_table_ptr = device_table::create(1, &col);
+   * 
+   * // Because `device_table` is allocated with managed memory, a pointer to 
+   * // the object can be passed directly into device code
+   * some_kernel<<<...>>>(device_table_ptr.get());
+   * ```
+   *
+   * @param num_columns The number of columns
+   * @param cols Array of columns
+   * @return A unique_ptr containing a device_table object
+   *---------------------------------------------------------------------------**/
+  static auto create(gdf_size_type num_columns, gdf_column* cols[],
+                     cudaStream_t stream = 0) {
+    auto deleter = [](device_table* d) { d->destroy(); };
+
+    std::unique_ptr<device_table, decltype(deleter)> p{
+        new device_table(num_columns, cols, stream), deleter};
+
+    int dev_id = 0;
+    CUDA_TRY(cudaGetDevice(&dev_id));
+    CUDA_TRY(cudaMemPrefetchAsync(p.get(), sizeof(*p), dev_id, stream))
+
+    CHECK_STREAM(stream);
+
+    return p;
+  }
+
+  static auto create(cudf::table& t, cudaStream_t stream = 0) {
+    return device_table::create(t.num_columns(), t.begin(), stream);
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Destroys the `device_table`.
+   *
+   * @note This function is required because the destructor is protected to prohibit
+   * stack allocation.
+   *
+   *---------------------------------------------------------------------------**/
+  void destroy(void){
+      delete this;
+  }
+
+  device_table() = delete;
+  device_table(device_table const& other) = delete;
+  device_table& operator=(device_table const& other) = delete;
+
+  /** 
+   * @brief  Updates the size of the gdf_columns in the table
```

Comment:
Yeah, it's a lie.

It's only used for the output of groupby. 

The outputs of groupby are pre-allocated outside of the libcudf API. Since you can't know _a priori_ what the exact output size of a groupby is, they are allocated using the upper bound of the size of the input. 

This function is to update the size of the columns to the actual groupby output size, but doesn't reallocate memory. 

I think this issue goes away when we move to algorithm's allocating and returning their outputs.

Reply:
Okay, totally removed this.